libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/DialogManager.cs

```diff
@@ -227,21 +224,126 @@ public async Task<DialogManagerResult> OnTurnAsync(ITurnContext context, Cancell
+            var snapshot = GetActiveDialogContext(dc).State.GetMemorySnapshot();
+            var traceActivity = (Activity)Activity.CreateTraceActivity("BotState", "https://www.botframework.com/schemas/botState", snapshot, traceLabel);
+            await dc.Context.SendActivityAsync(traceActivity, cancellationToken).ConfigureAwait(false);
+        }
+
+        // We should only cancel the current dialog stack if the EoC activity is coming from a parent (a root bot or another skill).
+        // When the EoC is coming back from a child, we should just process that EoC normally through the 
+        // dialog stack and let the child dialogs handle that.
+        private static bool IsEocComingFromParent(ITurnContext turnContext)
+        {
+            // To determine the direction we check callerId property which is set to the parent bot
+            // by the BotFrameworkHttpClient on outgoing requests.
+            return !string.IsNullOrWhiteSpace(turnContext.Activity.CallerId);
+        }
+
+        // Recursively walk up the DC stack to find the active DC.
+        private static DialogContext GetActiveDialogContext(DialogContext dialogContext)
+        {
+            var child = dialogContext.Child;
+            if (child == null)
+            {
+                return dialogContext;
+            }
+
+            return GetActiveDialogContext(child);
+        }
+
+        private async Task<DialogTurnResult> HandleSkillOnTurnAsync(DialogContext dc, CancellationToken cancellationToken)
+        {
+            // the bot is running as a skill. 
+            var turnContext = dc.Context;
+
+            // Process remote cancellation
+            if (turnContext.Activity.Type == ActivityTypes.EndOfConversation && dc.ActiveDialog != null && IsEocComingFromParent(turnContext))
+            {
+                // Handle remote cancellation request from parent.
+                var activeDialogContext = GetActiveDialogContext(dc);
+
+                var remoteCancelText = "Skill was canceled through an EndOfConversation activity from the parent.";
+                await turnContext.TraceActivityAsync($"{typeof(Dialog).Name}.RunAsync()", label: $"{remoteCancelText}", cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                // Send cancellation message to the top dialog in the stack to ensure all the parents are canceled in the right order. 
+                return await activeDialogContext.CancelAllDialogsAsync(true, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            // Handle reprompt
+            // Process a reprompt event sent from the parent.
+            if (turnContext.Activity.Type == ActivityTypes.Event && turnContext.Activity.Name == DialogEvents.RepromptDialog)
             {
-                snapshotDc = snapshotDc.Child;
+                await dc.RepromptDialogAsync(cancellationToken).ConfigureAwait(false);
+
+                // TODO: check with steve what we return here.
```

Comment:
You should check your DC to see if there's anything on the stack. If not return Empty otherwise return Waiting.

Reply:
Added check and test for handling reprompts when there is no active dialog.